File: Assignment 1/task2.scala
import org.apache.spark.{SparkConf, SparkContext, HashPartitioner}
import org.json4s._
import org.json4s.jackson.JsonMethods._
import java.io._

object task2 {
  implicit val formats: DefaultFormats.type = DefaultFormats

  def main(args: Array[String]): Unit = {
    // Expected args: <inputFile> <outputFile> <numPartitions>
    if (args.length != 3) {
      println("Usage: spark-submit --class Task2 <jar_file> <input_filepath> <output_filepath> <n_partitions>")
      sys.exit(1)
    }

    val inputPath     = args(0)
    val outputPath    = args(1)
    val numPartitions = args(2).toInt

    val conf = new SparkConf().setAppName("Task2").setMaster("local[*]")
    val sc   = new SparkContext(conf)
    sc.setLogLevel("ERROR")

    try {
      // 1) Read the input with 20 partitions to closely match PySpark’s default distribution
      val rawRDD = sc.textFile(inputPath, 20)

      // 2) Safely parse JSON and extract "business_id"
      val businessRDD = rawRDD.flatMap { line =>
        parseOpt(line)
      }.flatMap { jval =>
        (jval \ "business_id").extractOpt[String]
      }.cache()

      // ---------------------------
      // DEFAULT PARTITION ANALYSIS
      // ---------------------------
      val defaultStart = System.currentTimeMillis()

      // We simply reuse the RDD as-is, which already has 20 partitions
      val defaultTop10 = businessRDD
        .map(bid => (bid, 1))
        .reduceByKey(_ + _)
        .takeOrdered(10)(Ordering.by { case (id, count) => (-count, id) })

      val defaultTimeSecs =
        (System.currentTimeMillis() - defaultStart) / 1000.0

      val defaultNumPartitions = businessRDD.getNumPartitions
      val defaultPartitionSizes = businessRDD
        .mapPartitions(iter => Iterator(iter.size))
        .collect()

      // ---------------------------
      // CUSTOM PARTITION ANALYSIS
      // ---------------------------
      val customStart = System.currentTimeMillis()

      // Partition by a HashPartitioner using user-specified numPartitions
      val partitionedRDD = businessRDD
        .map(bid => (bid, 1))
        .partitionBy(new HashPartitioner(numPartitions))
        .cache()

      val customTop10 = partitionedRDD
        .reduceByKey(_ + _)
        .takeOrdered(10)(Ordering.by { case (id, count) => (-count, id) })

      val customTimeSecs =
        (System.currentTimeMillis() - customStart) / 1000.0

      val customNumPartitions = partitionedRDD.getNumPartitions
      val customPartitionSizes = partitionedRDD
        .mapPartitions(iter => Iterator(iter.size))
        .collect()

      // ---------------------------
      // Build JSON Output
      // ---------------------------
      val jsonOutput =
        s"""
           |{
           |  "default": {
           |    "n_partition": $defaultNumPartitions,
           |    "n_items": ${defaultPartitionSizes.mkString("[", ", ", "]")},
           |    "exe_time": $defaultTimeSecs
           |  },
           |  "customized": {
           |    "n_partition": $customNumPartitions,
           |    "n_items": ${customPartitionSizes.mkString("[", ", ", "]")},
           |    "exe_time": $customTimeSecs
           |  }
           |}
           |""".stripMargin

      val writer = new PrintWriter(new File(outputPath))
      writer.write(jsonOutput)
      writer.close()

    } finally {
      sc.stop()
    }
  }
}

File: Assignment 1/task2.py
import sys
import json
import time
from pyspark import SparkContext

def process_partition(iterator):
    """
    Processes a partition and counts occurrences of business_id.
    """
    count_map = {}
    for business_id, _ in iterator:
        count_map[business_id] = count_map.get(business_id, 0) + 1
    return iter(count_map.items())

def analyze_partitioning(input_path, output_path, num_partitions):
    """
    Analyze the effects of default and custom partitioning on business reviews data.
    """

    # Step 1: Initialize Spark Context
    sc = SparkContext("local[*]", "DSCI 553 Task 2")
    sc.setLogLevel("ERROR")

    try:
        # Step 2: Load dataset and parse JSON records
        review_rdd = sc.textFile(input_path).map(json.loads)

        # Step 3: Default partitioning analysis
        start_time = time.time()
        default_top10 = (review_rdd.map(lambda x: (x["business_id"], 1))
                         .reduceByKey(lambda a, b: a + b)
                         .takeOrdered(10, key=lambda x: (-x[1], x[0])))
        default_time = time.time() - start_time

        default_partitions = review_rdd.getNumPartitions()
        default_partition_sizes = review_rdd.glom().map(len).collect()

        # Step 4: Custom partitioning analysis
        start_time = time.time()
        partitioned_rdd = (review_rdd.map(lambda x: (x["business_id"], 1))
                           .partitionBy(num_partitions, lambda key: hash(key) % num_partitions)  # Fix applied
                           .mapPartitions(process_partition)
                           .cache())

        custom_top10 = (partitioned_rdd.reduceByKey(lambda a, b: a + b)
                        .takeOrdered(10, key=lambda x: (-x[1], x[0])))
        custom_time = time.time() - start_time

        custom_partitions = partitioned_rdd.getNumPartitions()
        custom_partition_sizes = partitioned_rdd.glom().map(len).collect()

        # Step 5: Save output to JSON file
        output_data = {
            "default": {"n_partition": default_partitions, "n_items": default_partition_sizes, "exe_time": default_time},
            "customized": {"n_partition": custom_partitions, "n_items": custom_partition_sizes, "exe_time": custom_time}
        }

        with open(output_path, "w") as f:
            json.dump(output_data, f, indent=4)

    finally:
        sc.stop()

if __name__ == "__main__":
    if len(sys.argv) != 4:
        print("Usage: spark-submit task2.py <input_filepath> <output_filepath> <n_partition>")
        sys.exit(1)

    analyze_partitioning(sys.argv[1], sys.argv[2], int(sys.argv[3]))

























# import sys
# import json
# import time
# from pyspark import SparkContext


# def partition_function(key):
#     """
#     Custom partition function using hash-based partitioning.
#     This helps distribute data more evenly across partitions.
#     """
#     return hash(key) % num_partitions


# def process_partition(iterator):
#     """
#     Helper function to count occurrences of business_id within a partition.
#     """
#     count_map = {}
#     for business_id, _ in iterator:
#         count_map[business_id] = count_map.get(business_id, 0) + 1
#     return iter(count_map.items())


# def analyze_partitioning(input_path, output_path, num_partitions):
#     """
#     Function to analyze default and custom partitioning on business reviews data.
#     Measures execution time and data distribution across partitions.
#     """
#     sc = SparkContext("local[*]", "DSCI 553 Task 2")
#     sc.setLogLevel("ERROR")

#     try:
#         # Load input data
#         review_rdd = sc.textFile(input_path).map(json.loads)

#         # Default partitioning analysis
#         start_time = time.time()
#         top10_default = (review_rdd.map(lambda x: (x["business_id"], 1))
#                          .reduceByKey(lambda a, b: a + b)
#                          .takeOrdered(10, key=lambda x: (-x[1], x[0])))
#         default_time = time.time() - start_time

#         default_partitions = review_rdd.getNumPartitions()
#         default_partition_sizes = review_rdd.glom().map(len).collect()

#         # Custom partitioning analysis
#         start_time = time.time()
#         partitioned_rdd = (review_rdd.map(lambda x: (x["business_id"], 1))
#                            .partitionBy(num_partitions, partition_function)
#                            .mapPartitions(process_partition))

#         top10_custom = (partitioned_rdd.reduceByKey(lambda a, b: a + b)
#                         .takeOrdered(10, key=lambda x: (-x[1], x[0])))
#         custom_time = time.time() - start_time

#         custom_partitions = partitioned_rdd.getNumPartitions()
#         custom_partition_sizes = partitioned_rdd.glom().map(len).collect()

#         # Prepare output in the required format
#         output_data = {
#             "default": {
#                 "n_partition": default_partitions,
#                 "n_items": default_partition_sizes,
#                 "exe_time": default_time
#             },
#             "customized": {
#                 "n_partition": custom_partitions,
#                 "n_items": custom_partition_sizes,
#                 "exe_time": custom_time
#             }
#         }

#         # Write output to JSON file
#         with open(output_path, "w") as f:
#             json.dump(output_data, f, indent=4)

#         print(f"✅ Output successfully written to {output_path}")

#     except Exception as e:
#         print(f"❌ ERROR: {str(e)}")
#         sys.exit(1)
    
#     finally:
#         sc.stop()


# if __name__ == "__main__":
#     if len(sys.argv) != 4:
#         print("Usage: spark-submit task2.py <input_filepath> <output_filepath> <n_partition>")
#         sys.exit(1)

#     review_filepath = sys.argv[1]
#     output_filepath = sys.argv[2]
#     num_partitions = int(sys.argv[3])

#     analyze_partitioning(review_filepath, output_filepath, num_partitions)
